core/server/src/main/java/alluxio/master/journal/ProtoBufJournalFormatter.java

```diff
@@ -40,15 +48,35 @@ public void serialize(JournalEntry entry, OutputStream outputStream) throws IOEx
   @Override
   public JournalInputStream deserialize(final InputStream inputStream) throws IOException {
     return new JournalInputStream() {
+      private final byte[] mBuffer = new byte[1024];
       private long mLatestSequenceNumber;
 
       @Override
       public JournalEntry getNextEntry() throws IOException {
-        JournalEntry entry = JournalEntry.parseDelimitedFrom(inputStream);
-        if (entry != null) {
-          mLatestSequenceNumber = entry.getSequenceNumber();
+        try {
+          int firstByte = inputStream.read();
+          if (firstByte == -1) {
+            return null;
+          }
+          int size = CodedInputStream.readRawVarint32(firstByte, inputStream);
+          byte[] buffer = size <= mBuffer.length ? mBuffer : new byte[size];
+          int bytes = inputStream.read(buffer, 0, size);
+          if (bytes < size) {
+            throw new InvalidProtocolBufferException(
+                "Journal entry was truncated. Expected to read " + size + " bytes but only got "
+                    + bytes);
+          }
+          JournalEntry entry = JournalEntry.parseFrom(new ByteArrayInputStream(buffer, 0, size));
+          if (entry != null) {
+            mLatestSequenceNumber = entry.getSequenceNumber();
+          }
+          return entry;
+        } catch (InvalidProtocolBufferException e) {
+          LOG.warn("Failed to read journal entry", e);
+          // Barring IO corruption, this means that the master crashed while writing the last
+          // journal entry, so we can ignore this last entry.
+          return null;
```

Comment:
I think it depends whether we can have parallel outstanding requests being written to the journal at the same time in Alluxio? TL;DR. In other systems, I've seen bugs that look like corruptions because we write to two journal locations in parallel. 

At the time of pulling "the plug" the latter write finished, but the former hadn't. After running the consistency checker, we thought we had a bogus NVRAM card (corrupted the former log entry). That was incorrect though because those writes were "in-flight" and hadn't been acked to the client yet. However, that was a strongly consistent system. The solution we implemented was to keep track of the last ack'd entries and only replay up until that point.